### ISSUE TITLE
Canvas does not work AT ALL outside of constructor

### ISSUE THREAD

**Issue by Andrew-lll** [dotnet/maui#8142]
### Description

See https://docs.microsoft.com/en-us/dotnet/maui/user-interface/graphics/draw

_All_ drawing examples work **_only_** in the canvas elements (IDrawable) constructor Draw(ICanvas, RectF), not outside/after (such as on an event or a button click).  Because of this, the canvas is unusable.

### All canvas drawing functions fail when connected to any event or button press, see error below:

![Canvas Error](https://user-images.githubusercontent.com/19860489/174386734-50702cce-661e-4433-a6c2-8a33066858c0.png)

### Steps to Reproduce

See complete code example above.  Will provide above example code files if needed.

### Version with bug

6.0.400 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Windows

### Affected platform versions

All platforms, all versions

**Comment by janseris** [comment]
I also had large problems getting this usable. In the end I went for Blazor. 
It seems like the IDrawable object must be a central entity which violates single responsibility principle.

**Comment by Andrew-lll** [comment]
Others have the same problem:

**Cant draw using mouse event:** https://stackoverflow.com/questions/71715884/idrawable-mouse-event

@janseris Yes, I implement the Draw function, which is called automatically when the IDrawable is created - and as you mention, I can draw inside it - **but only once**.  I also have a functioning js canvas (because the web canvas api works & has documentation).  The problem is, after initialization of the canvas, it is disposed of, and there is no documentation how to continue drawing on it or how to re-call the Draw constructor.

**Comment by pierre01** [comment]
Call **Invalidate** on your GraphicsView control (Canvas) it will call the Draw method automatically on your IDrawable.
But Keep the drawing of your points in the Draw Method.

It should work.

**Comment by janseris** [comment]
@pierre01 the problem is defining the draw method content dynamically, e. g. draw shapes from API. Drawing hardcoded shapes is simple just as you are showing.

**Comment by pierre01** [comment]
@janseris 
 Please look at the samples I give, one of them is drawing random size lines based on code. Not hardcoded.

**Comment by Andrew-lll** [comment]
@pierre01 Thank you for the examples - with them I implemented:

### The temporary workaround:
(1) Have a button hooked up to a function that calls Canvas.Invalidate() (which implicitly calls the Draw() function again).
(2) Updated the code inside the Draw() function to redraw a complete scene from scratch based on some data (instead of being additive)

**As you can see above, my issue now is that Invalidate() + the implied Draw() constructor wipes the entire canvas.**  This might make sense in some scenarios (redrawing a scene in a game at 60fps), but it doesn't make any sense in other scenarios (for example when a user is drawing lines one after another).

### Some suggestions for the Maui team:

**1. Documentation:** This behavior needs to be explained in the Maui docs.  Explain Invalidating, explain that all drawing code must be in the draw constructor, reference Skia docs that explain this.  
**2. Actually change this behavior:** I should be able to (in a simple way) draw a Line on the canvas, and then at a later time draw another line (without my previous line + everything else disappearing).  It makes sense to be able to Invalidate and clear the canvas at times, but it doesn't makes sense to HAVE to do that, and it also doesn't make sense that all drawing code MUST be in the Draw() function (especially when you already have references to the Canvas and Drawable).

See https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API

**Comment by adegeo** [comment]
Hi @Andrew-lll can you explain more about (1) related to documentation? What do you mean by drawing in the constructor?

My general thoughts on Skia so far are that it's not good for games. Which is unfortunate, I personally find it overengineered and too complex to build a game framework on. Once someone actually does that, creating a gameframework that is built directly for Skia, you may have better luck doing something game-related. Skia itself isn't built for that, even the github for Google's repository only says `Skia is a complete 2D graphic library for drawing Text, Geometries, and Images.` and obviously doesn't mention games or anything like that at all, so it's not a goal of the engineering design of Skia.

For (2) as a workaround you could look at using an offscreen buffer, draw/copy all the results at the end, to a temp bitmap. Each refresh of the graphics view via invalidate would first draw the bitmap image and then run the code to draw on top of it, basically  a fake preserve... I would file a new feature request about adding a property or something to disable the clear on the canvas when the graphics view is invalidated. If you don't create a specific issue about this, I fear this enhancement request may get lost as it's been tacked on to this issue.

**Comment by mattleibow** [comment]
@Andrew-lll you observation is a design choice. The canvas only exists during the draw loop. If you wish to update from the outside, the usual way is to update some properties or data, and then invalidate the graphics view.

The samples by @pierre01 https://github.com/dotnet/maui/issues/8142#issuecomment-1159372990 are pretty great and show ways of doing things, and @adegeo also has some good suggestions: https://github.com/dotnet/maui/issues/8142#issuecomment-1180787732

Basically, this is how most native platforms work and for the support of arbitrary drawing operations, you may wish to draw on an off screen image. And again, @adegeo is correct, please create a new issue on this freature.

Maui by default has an immediate mode graphics system, you are looking for a retained mode. This does exist, but you will be using the Line and Polygon shapes:

https://learn.microsoft.com/dotnet/maui/user-interface/controls/shapes
